File: v4-client-js/examples/transfer_example_withdraw_other.ts
import { EncodeObject } from '@cosmjs/proto-signing';
import { Method } from '@cosmjs/tendermint-rpc';
import Long from 'long';

import { TEST_RECIPIENT_ADDRESS } from '../__tests__/helpers/constants';
import { BECH32_PREFIX } from '../src';
import { Network } from '../src/clients/constants';
import LocalWallet from '../src/clients/modules/local-wallet';
import { SubaccountInfo } from '../src/clients/subaccount';
import { ValidatorClient } from '../src/clients/validator-client';
import { DYDX_TEST_MNEMONIC } from './constants';

async function test(): Promise<void> {
  const wallet = await LocalWallet.fromMnemonic(DYDX_TEST_MNEMONIC, BECH32_PREFIX);
  console.log(wallet);

  const client = await ValidatorClient.connect(Network.testnet().validatorConfig);
  console.log('**Client**');
  console.log(client);

  const subaccount = new SubaccountInfo(wallet, 0);

  const amount = new Long(100_000_000);

  const msgs: Promise<EncodeObject[]> = new Promise((resolve) => {
    const msg = client.post.composer.composeMsgWithdrawFromSubaccount(
      subaccount.address,
      subaccount.subaccountNumber,
      0,
      amount,
      TEST_RECIPIENT_ADDRESS,
    );

    resolve([msg]);
  });

  const totalFee = await client.post.simulate(subaccount.wallet, () => msgs, undefined);
  console.log('**Total Fee**');
  console.log(totalFee);

  const amountAfterFee = amount.sub(Long.fromString(totalFee.amount[0].amount));
  console.log('**Amount after fee**');
  console.log(amountAfterFee);

  const tx = await client.post.withdraw(
    subaccount,
    0,
    amountAfterFee,
    TEST_RECIPIENT_ADDRESS,
    Method.BroadcastTxCommit,
  );
  console.log('**Withdraw and Send**');
  console.log(tx);
}

test()
  .then(() => {})
  .catch((error) => {
    console.log(error.message);
  });

File: v4-client-py-v2/examples/transfer_example_withdraw_other.py
import asyncio
from functools import partial

from dydx_v4_client.network import TESTNET
from dydx_v4_client.node.client import NodeClient
from dydx_v4_client.node.message import subaccount, withdraw
from dydx_v4_client.wallet import Wallet
from tests.conftest import DYDX_TEST_MNEMONIC, TEST_ADDRESS


async def test():
    node = await NodeClient.connect(TESTNET.node)

    wallet = await Wallet.from_mnemonic(node, DYDX_TEST_MNEMONIC, TEST_ADDRESS)

    withdraw_tx = partial(withdraw, subaccount(TEST_ADDRESS, 0), TEST_ADDRESS, 0)
    amount = 100_000_000

    simulated_txs = node.builder.build(wallet, withdraw_tx(amount))
    simulation = await node.simulate(simulated_txs)
    print("**Simulate**")
    print(simulation)

    fee = node.calculate_fee(simulation.gas_info.gas_used)
    print("**Total Fee**")
    print(fee)

    final_withdraw_tx = withdraw_tx(amount - fee.amount[0].amount)
    final_txs = node.build(wallet, final_withdraw_tx, fee)
    response = await node.broadcast(final_txs)
    print("**Withdraw and Send**")
    print(response)


asyncio.run(test())
